{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Ch2 sequence analysis**\n",
    "\n",
    "* in preliminary CH1 analyses, I did some sequence trimming in conda locally. It is more streamlined to do trimming in qiime\n",
    "* manifest file directs qiime to find raw files in RIS storage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* activate Qiime2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## in WashU RIS cluster: see RIS document for context but examples below)\n",
    "\n",
    "bsub -Is -G compute-rachel.penczykowski -q general-interactive -R 'rusage[mem=16GB]' -a 'docker(quay.io/qiime2/core:2023.2)' /bin/bash\n",
    "bsub -Is -G compute-rachel.penczykowski -q general-interactive -R 'rusage[mem=32GB]' -a 'docker(quay.io/qiime2/core:2023.2)' /bin/bash\n",
    "bsub -Is -G compute-rachel.penczykowski -q general-interactive -R 'rusage[mem=128GB]' -a 'docker(quay.io/qiime2/core:2023.2)' /bin/bash"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* change directory to files location\n",
    "* my manifest file directs qiime to find raw files in RIS storage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "cd /storage1/fs1/rachel.penczykowski/Active/ptanford/CH2/ #put manifest file in rs1 folder\n",
    "\n",
    "qiime tools import \\\n",
    "--type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "--input-path CH2_manifest.txt \\\n",
    "--output-path raw_seqs.qza \\\n",
    "--input-format PairedEndFastqManifestPhred33V2 \n",
    "\n",
    "## demux summarize (notes on why we do this to hopefully follow but at a minimum sanity check of import success)\n",
    "## I believe the main point of this is to have sequence statistics as data quality etc. checks visualizable at\n",
    "## view.qiime2.org\n",
    "\n",
    "qiime demux summarize \\\n",
    "--i-data raw_seqs.qza \\\n",
    "--o-visualization raw_seqs.qzv "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**REMOVING PRIMERS + adapters and DISPOSING OF NON-PRIMER READS **\n",
    "**I think I can remove primer + adapter together. For CH1 it was important to separate fungi and bacteria**\n",
    "\n",
    "* fITS7 GTGAATCATCGAATCTTTG (Ihrmark et al. 2012) \n",
    "* ITS4 TCCTCCGCTTATTGATATGC (White et al. 1990)\n",
    "\n",
    "* Forward overhang: TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG\n",
    "* Reverse overhang: GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG\n",
    "\n",
    "* adfITS7 TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGTGAATCATCGAATCTTTG\n",
    "* adITS4 GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGTCCTCCGCTTATTGATATGC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## remove primers (adapter overhang + ITS primer seq) with cutadapt\n",
    "\n",
    "qiime cutadapt trim-paired \\\n",
    "--i-demultiplexed-sequences raw_seqs.qza \\\n",
    "--p-front-f TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGTGAATCATCGAATCTTTG \\\n",
    "--p-front-r GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGTCCTCCGCTTATTGATATGC \\\n",
    "--p-match-read-wildcards \\\n",
    "--p-match-adapter-wildcards \\\n",
    "--verbose \\\n",
    "--o-trimmed-sequences raw_seqs_trim1.qza\n",
    "\n",
    "## assess files post trim\n",
    "qiime demux summarize \\\n",
    "--i-data raw_seqs_trim1.qza \\\n",
    "--o-visualization trimmed_demuxsum.qzv "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Run dada2 for denoising**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## paired maxEE 2,2\n",
    "qiime dada2 denoise-paired \\\n",
    "--i-demultiplexed-seqs raw_seqs_trim1.qza \\\n",
    "--p-trunc-len-f 0 \\\n",
    "--p-trunc-len-r 10 \\\n",
    "--p-max-ee-f 2 \\\n",
    "--p-max-ee-r 2 \\\n",
    "--o-table dada2_table_22.qza \\\n",
    "--o-denoising-stats maxEE2-2/dada2_stats_22.qza \\\n",
    "--o-representative-sequences maxEE2-2/RepSeqs22.qza \\\n",
    "--verbose\n",
    "\n",
    "## paired no maxEE - no reads merge. will have to go with single\n",
    "qiime dada2 denoise-paired \\\n",
    "--i-demultiplexed-seqs raw_seqs_trim1.qza \\\n",
    "--p-trunc-len-f 0 \\\n",
    "--p-trunc-len-r 10 \\\n",
    "--o-table dada2_table_noEE.qza \\\n",
    "--o-denoising-stats dada2_stats_noEE.qza \\\n",
    "--o-representative-sequences RepSeqs_noEE.qza \\\n",
    "--verbose\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## single maxEE 2 (so only forward reads)\n",
    "qiime dada2 denoise-single \\\n",
    "--i-demultiplexed-seqs raw_seqs_trim1.qza \\\n",
    "--p-max-ee 2 \\\n",
    "--p-trunc-len 0 \\\n",
    "--o-table dada2_table_2.qza \\\n",
    "--o-denoising-stats dada2_stats_2.qza \\\n",
    "--o-representative-sequences RepSeqs_2.qza \\\n",
    "--verbose\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "qiime feature-table summarize \\\n",
    "  --i-table dada2_table_2.qza \\\n",
    "  --o-visualization feature_table_2.qzv \\\n",
    "  --m-sample-metadata-file CH2_metadata.txt\n",
    "  \n",
    "qiime feature-table tabulate-seqs \\\n",
    "--i-data RepSeqs_2.qza \\\n",
    "--o-visualization RepSeqs_2.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Classify taxa with unite database classifier**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## copying code I used to make prior fungal classifier. Will use it again to classify these taxa but I am not running this anew\n",
    "## the below only ran on the cluster with 128GB RAM. 64 was not enough\n",
    "qiime feature-classifier fit-classifier-naive-bayes \\\n",
    " --i-reference-reads unite-refs-ver9_99_25.07.2023.qza \\\n",
    " --i-reference-taxonomy unite-taxonomy-ver9_99_25.07.2023.qza \\\n",
    " --o-classifier unite-2023-classifier.qza\n",
    "\n",
    "\n",
    "## classify my samples\n",
    "## this doesn't run on 64GB\n",
    "qiime feature-classifier classify-sklearn \\\n",
    "  --i-classifier unite-2023-classifier.qza \\\n",
    "  --i-reads RepSeqs_2.qza \\\n",
    "  --o-classification Taxonomy_2.qza\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sanity check - diversity plots (prior to phylogenetic IDs) in qiime2**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences RepSeqs_2.qza \\\n",
    "  --o-alignment testAlignedRepSeqs.qza \\\n",
    "  --o-masked-alignment testMaskedRepSeqs.qza \\\n",
    "  --o-tree testUnrooted-tree.qza \\\n",
    "  --o-rooted-tree testRooted-tree.qza\n",
    "\n",
    "qiime diversity core-metrics-phylogenetic \\\n",
    "  --i-phylogeny testRooted-tree.qza \\\n",
    "  --i-table dada2_table_2.qza \\\n",
    "  --p-sampling-depth 10000 \\\n",
    "  --m-metadata-file CH2_metadata.txt \\\n",
    "  --output-dir core-metrics-results\n",
    "\n",
    "## plots look overall good!! samples clustering by soil type. much more spread \n",
    "## of live samples with jaccard - unique fungal spp in live pots\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Troubleshooting qza to phyloseq issues - possibly due to my sample names being numbers. Per this:\n",
    "https://forum.qiime2.org/t/change-sample-ids-after-running-dada2/3918\n",
    "I can use the feaature table group function to direct qiime to new sample IDs\n",
    "\n",
    "\"Luckily, I have just the trick for you! Take a look at feature-table group! That might sound weird at first, but, stick with me. What you can do is add a new column to your metadata file that contains the new sample IDs. Then, when you run the group command, point the --m-metadata-column to the column containing your new sample IDs. The --p-mode parameter don't matter (you can pick whatever you want) - because you are \"grouping\" the values into the same \"groups\", just with new names, none of the abundances will change. Et voila! Moving forward, just drop the old sample ID column from your metadata and replace it with the new sample ID column. :recycle:\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "feature table group \\\n",
    "--i-table \\\n",
    "--p-axis [feature|sample] \\\n",
    "--m-metadata-file MULTIPLE PATH \\\n",
    "--m-metadata-column MetadataColumn[Categorical] \\\n",
    "--p-mode [median-ceiling|sum|mean-ceiling] \\ \n",
    "--o-grouped-table ARTIFACT PATH FeatureTable[Frequency]\n",
    "\n",
    "#p-mode is required but arbitrary\n",
    "\n",
    "qiime feature-table group \\\n",
    "--i-table dada2_table_2.qza \\\n",
    "--p-axis sample \\\n",
    "--m-metadata-file CH2_metadata_newIDs.txt \\\n",
    "--m-metadata-column sample \\\n",
    "--p-mode sum \\\n",
    "--o-grouped-table dada2_table_2_newIDs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above ran! I then had to:\n",
    "* Save a new keemi-approved metadata sheet that replaced the prior sample-id names (101, 102...., under the #q2:types column) with sample (Sample101, Sample102...), deleting the sample-id column entirely. The \"CH2_metadata_newIDs\" file has both sample types in order to run the grouping code and output a new feature table with the new ids. \"Ch2_metadata_newIDs2\" (note the 2 at the end) is the new metadata file associated with the new dada2 table, for use for importing into R with qza to phyloseq :)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
